src/relay/backend/contrib/cmsisnn/relay_to_tir.cc

```diff
@@ -0,0 +1,140 @@
+      num_rows *= shape[i].as<tir::IntImmNode>()->value;
+    }
+
+    // calculate multiplier and shift for CMSIS-NN softmax API
+    // Note: TensorFlow Lite Micro assumptions
+    // Output zero point and scale are fixed to -128 and 1 / 256
+    // https://github.com/tensorflow/tflite-micro/blob/d97cd0908d8cf5021e9d86f05a49888bee28c2a4/tensorflow/lite/micro/kernels/softmax_common.cc#L47
+    double beta = 1.0;
+    int32_t input_bits = 5;
+    double beta_multiplier = (beta * quant_scale * (1 << (31 - input_bits)));
+    beta_multiplier = std::min<double>(beta_multiplier, (1ll << 31) - 1.0);
+    auto mult_shift_pair = tvm::relay::qnn::GetFixedPointMultiplierShift(beta_multiplier);
+    int32_t mult = std::get<0>(mult_shift_pair);
+    int32_t shift = std::get<1>(mult_shift_pair);
+    int32_t diff_min = (1 << 5) - 1;
+    diff_min <<= (31 - 5);
+    diff_min >>= shift;
+    diff_min *= -1;
+
+    auto in_var = tir::Var("input", DataType::Handle(8));
+    auto out_var = tir::Var("output", DataType::Handle(8));
+
+    Array<tir::Var> func_signature{in_var, out_var};
+
+    tvm::Array<PrimExpr> args = {
+        tir::StringImm("arm_softmax_s8"),    in_var,
+        IntImm(DataType::Int(32), num_rows), IntImm(DataType::Int(32), row_size),
+        IntImm(DataType::Int(32), mult),     IntImm(DataType::Int(32), shift),
+        IntImm(DataType::Int(32), diff_min), out_var};
+    tir::Stmt body =
+        tir::Evaluate(tvm::tir::Call(DataType::Int(8), tir::builtin::call_extern(), args));
+
+    Map<String, ObjectRef> dict_attrs;
+    dict_attrs.Set("global_symbol", func_name_);
+    dict_attrs.Set("tir.noalias", Bool(true));
+
+    primfunc_ = tir::PrimFunc(func_signature, body, VoidType(), Map<tir::Var, tir::Buffer>(),
+                              DictAttrs(dict_attrs));
+  }
+
+  void VisitExpr_(const CallNode* call) final {
+    auto* func = call->op.as<FunctionNode>();
+    if (func == nullptr) {
+      return;
+    }
+
+    auto comp_name = func->GetAttr<String>(attr::kComposite);
+    if (comp_name.defined() && comp_name == "cmsisnn.quantized_softmax") {
+      emit_softmax_tir(func->body);
+    }
+  }
+
+ public:
+  String func_name_;
+  tir::PrimFunc primfunc_;
+};
+
+IRModule GenerateTIR(IRModule mod) {
+  String func_name;
+  Function func;
+
+  // Obtain external Relay Function that needs to be translated into TIR
+  ICHECK(mod->functions.size() == 1) << "Supports modules with single external Relay function.";
+  for (auto kv : mod->functions) {
+    func = Downcast<Function>(kv.second);
+    func_name = func->GetAttr<String>(tvm::attr::kGlobalSymbol).value();
+  }
+
+  // Prepare PrimFunc from Relay Function
+  auto relay_to_tir = RelayToTIR(func_name);
+  relay_to_tir.VisitExpr(func->body);
+
+  // Build the TIR IRModule from the generated PrimFunc
+  Map<GlobalVar, BaseFunc> var_func_map;
+  var_func_map.Set(GlobalVar(func_name), relay_to_tir.primfunc_);
+  return IRModule(var_func_map);
+}
+
+transform::Pass RelayToTIR() {
```

Comment:
does this need updating now that #8423 is merged?

Reply:
Unfortunately not, as CMSIS NN currently requires `TIRToRuntime` as well to generate the structures. I don't think we should intertwine the two pieces of work if possible and unpick it later.